beater/pub.go

```diff
@@ -50,34 +47,31 @@ func newPublisher(pipeline beat.Pipeline, N int) (*publisher, error) {
 
 	p := &publisher{
 		client: client,
-
+		done:   make(chan struct{}),
 		// Set channel size to N - 1. One request will be actively processed by the
 		// worker, while the other concurrent requests will be buffered in the queue.
 		events: make(chan []beat.Event, N-1),
 	}
 
-	p.wg.Add(1)
 	go p.run()
 	return p, nil
 }
 
 // Stop closes all channels and waits for the the worker to stop.
 // The worker will drain the queue on shutdown, but no more events
 // will be published.
-func (p *publisher) Stop() {
-	p.drain()
+func (p *publisher) Stop(d time.Duration) {
+	close(p.done)
+	p.drain(d)
 	close(p.events)
 	p.client.Close()
-	p.wg.Wait()
 }
 
-// drain signals the publisher to stop accepting event batches, blocking
-// execution until all the queued events have been published.
-func (p *publisher) drain() {
-	p.m.Lock()
-	p.stopped = true
-	p.m.Unlock()
-	for {
+// drain waits until the events queue is empty or apm-server.shutdown_timeout seconds have elapsed
+func (p *publisher) drain(d time.Duration) {
```

Comment:
I think we don't need to check for the queue length at all, as we use the `beat.Client`. If we add the `bt.config.timeout` to the `beat.ClientConfig` in the `newPublisher` then the proper draining of the queue before shutting down should be handled by the beats framework, like:

```
func newPublisher(pipeline beat.Pipeline, N int, timeout time.Duration) (*publisher, error) {
        ...

	client, err := pipeline.ConnectWith(beat.ClientConfig{
		PublishMode: beat.GuaranteedSend,

		// TODO: We want to wait for events in pipeline on shutdown?
		//       If set >0 `Close` will block for the duration or until pipeline is empty
		WaitClose: timeout,
	})
	        ...
}
```

Reply:
Interesting, didn't notice that. But putting the timeout there doesn't seem to give the expected result...

Reply:
I used this recently for shutting down the server and it always behaved as expected - what was the issue you run into with it?

Reply:
yeah, the timeout is effective, but within this PR it panics. 
this is the simplest approach i came up with that seeems to work; is there any other way to close the server properly that doesn't add more timeouts, locks and code in general?

(i actually don't even know what is the point is the shutdown timeout. for what reasons might be the apm server stopped? does it matter if we have drain the queue or not? do we expect users to stop their agents before stopping the server?)

Reply:
I didn't try it in the context of this PR, but in addition what we currently have in master (and removing manually draining the queue).

Reply:
There are two different queues in play here. APM Servers "own" queue `p.events` and the libbeat's queue. `WaitClose: timeout` affects only the libbeat queue.

One of the reasons for draining the queue is when you want to do a rolling upgrade to your APM Servers. The process is like this:

1. You take the first APM Server out of the load balancer so it doesn't get any more requests. 
1. You signal the APM Server to shut down when it's finished delivering all the data it has queued up. If it takes longer than some timeout, we want it to just drop the data and exit.
1. You restart the APM Server, put it back in the load balancer
1. Do it to the next APM Server